pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,307 @@
+	// associated with the binding.
+	//
+	// 3. For every workload object associated with one or more statuscollectors, a
+	// combinedstatus resolution is created/updated.
+	// In the case of creation, the binding.Resolution's OwnerReference is used
+	// to identify the bindingpolicy that the resolution is associated with.
+	//
+	// The returned array contains:
+	//   - The identifiers of workload objects that should have their
+	//     associated workstatuses processed for evaluation.
+	//   - The identifiers of statuscollectors that need to be processed and
+	//     cached.
+	NoteBindingResolution(bindingResolution binding.Resolution, deleted bool) sets.Set[util.ObjectIdentifier]
+
+	// NoteStatusCollector notes a statuscollector's spec.
+	// The statuscollector is cached on the resolver's level, and is updated
+	// for every resolution it is involved with.
+	// The returned set contains the identifiers of workload objects that
+	// should have their associated workstatuses processed for evaluation.
+	// The statuscollector is assumed to be valid.
+	NoteStatusCollector(statusCollector *v1alpha1.StatusCollector) sets.Set[util.ObjectIdentifier]
+
+	// NoteWorkStatus notes a workstatus's content in the combinedstatus
+	// resolution associated with its source workload object.
+	// The returned boolean indicates whether the combinedstatus resolution
+	// associated with the workstatus was updated.
+	NoteWorkStatus(ctx context.Context, workStatusIdentifier util.ObjectIdentifier,
+		workStatusContent *runtime.RawExtension) bool
+
+	// ResolutionExists returns true if a combinedstatus resolution exists for
+	// the given binding name and workload object identifier.
+	ResolutionExists(bindingName string, objectIdentifier util.ObjectIdentifier) bool
+}
+
+// NewCombinedStatusResolver creates a new CombinedStatusResolver.
+func NewCombinedStatusResolver(celEvaluator *celEvaluator) (CombinedStatusResolver, error) {
+	celEvaluator, err := newCELEvaluator()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CEL evaluator: %w", err)
+	}
+
+	return &combinedStatusResolver{
+		celEvaluator:             celEvaluator,
+		bindingNameToResolutions: make(map[string]map[util.ObjectIdentifier]*combinedStatusResolution),
+	}, nil
+}
+
+type combinedStatusResolver struct {
+	sync.RWMutex
+	celEvaluator *celEvaluator
+	// bindingNameToResolutions is a map of binding names to their resolution
+	// entries. The latter is a map of object identifiers to their
+	// combinedstatus resolutions.
+	bindingNameToResolutions map[string]map[util.ObjectIdentifier]*combinedStatusResolution
+	// statusCollectorNameToClauses is a map of statuscollector names to their
+	// clauses. This serves as a cache that is the source of truth for
+	// statuscollectors that are used in the combinedstatus resolutions.
+	statusCollectorNameToClauses map[string]statusCollectorClauses
+}
+
+type statusCollectorClauses struct {
```

Comment:
This type looks like it has the same content as `v1alpha1.StatusCollectorSpec`, why does it exist?

Reply:
removed